// bespoken-jest-stare/lib/render/summary/TestSummary.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestSummary = void 0;
const Constants_1 = require("../Constants");
class TestSummary {
    static create(results) {
        const elements = [];
        const div = document.createElement("div");
        div.classList.add("my-3", "p-3", "bg-white", "rounded", "box-shadow", "summary");
        const h5 = document.createElement("h5");
        h5.classList.add("border-bottom", "pb-2", "display-5", "summary-title");
        h5.textContent = "Summary";
        div.appendChild(h5);
        div.id = "test-summary";
        elements.push(div);
        results.testResults.forEach((testResult, pathIndex) => {
            if (testResult.testResults == null) {
                console.error("Unexpected testResults field missing");
                if (testResult.assertionResults != null) {
                    console.warn("Attempting to use assertionResults: results are unpredictable");
                    testResult.testResults = testResult.assertionResults;
                }
            }
            const divMap = new Map();
            const testCountMap = new Map();
            const testCountArray = [];
            const testTitleDiv = document.createElement("div");
            testTitleDiv.classList.add("summary-test-suite");
            const testFileLink = document.createElement("a");
            const isFail = testResult.numFailingTests > 0;
            const testStatus = document.createElement("strong");
            testStatus.classList.add("summary-test-label");
            if (!isFail) {
                testStatus.classList.add("pass");
                testStatus.textContent = "PASS";
            }
            else {
                testStatus.classList.add("fail");
                testStatus.textContent = "FAIL";
            }
            const testFileLine = document.createElement("strong");
            testFileLine.classList.add("summary-test-label", "path");
            testFileLine.textContent = testResult.testFilePath;
            testFileLink.href = "#" + testResult.testFilePath;
            testFileLink.appendChild(testStatus);
            testFileLink.appendChild(testFileLine);
            testTitleDiv.appendChild(testFileLink);
            div.appendChild(testTitleDiv);
            testResult.testResults.forEach((test, testIndex) => {
                if (test.ancestorTitles.length > 0) {
                    test.ancestorTitles.forEach((title, index) => {
                        const titlesCopy = test.ancestorTitles.slice();
                        titlesCopy.splice(index + 1);
                        const key = titlesCopy.join(TestSummary.JOIN_CHAR);
                        if (!divMap.has(key)) {
                            const nestDiv = document.createElement("div");
                            nestDiv.classList.add("summary-ancestor-box");
                            nestDiv.id = key;
                            divMap.set(key, nestDiv);
                            if (index === 0) {
                                div.appendChild(nestDiv);
                            }
                            else {
                                titlesCopy.pop();
                                const parentKey = titlesCopy.join(TestSummary.JOIN_CHAR);
                                divMap.get(parentKey).appendChild(nestDiv);
                                const href = "#" + key + TestSummary.JOIN_CHAR
                                    + pathIndex + TestSummary.JOIN_CHAR + testIndex
                                    + TestSummary.JOIN_CHAR + index;
                                testCountMap.set(key, {
                                    locale: test.ancestorTitles[0],
                                    title,
                                    href,
                                    passingTests: 0,
                                    pendingTests: 0,
                                    failingTests: 0
                                });
                                testCountArray.push(key);
                            }
                        }
                        if (index !== 0) {
                            const testCount = testCountMap.get(key);
                            if (test.status === Constants_1.Constants.TEST_STATUS_PASS) {
                                testCount.passingTests = testCount.passingTests + 1;
                            }
                            if (test.status === Constants_1.Constants.TEST_STATUS_PEND) {
                                testCount.pendingTests = testCount.pendingTests + 1;
                            }
                            if (test.status === Constants_1.Constants.TEST_STATUS_FAIL) {
                                testCount.failingTests = testCount.failingTests + 1;
                            }
                            testCountMap.set(key, testCount);
                        }
                    });
                }
            });
            const locales = [...new Set(testCountArray.map((key) => testCountMap.get(key).locale))];
            testCountArray.forEach((key) => {
                const testCount = testCountMap.get(key);
                const testCountSpan = document.createElement("strong");
                testCountSpan.classList.add("summary-test-count");
                const testTotal = testCount.passingTests + testCount.pendingTests + testCount.failingTests;
                testCount.testTotal = testTotal;
                const testCountText = "[" + testCount.passingTests + "/" + testTotal + "]";
                testCountSpan.textContent = testCountText;
                const containerDiv = divMap.get(key);
                const ancestorLink = document.createElement("a");
                ancestorLink.textContent = locales.length > 1 ? `(${testCount.locale}) ${testCount.title}`
                    : testCount.title;
                ancestorLink.href = testCount.href;
                ancestorLink.classList.add("summary-test-label", "path");
                containerDiv.appendChild(TestSummary.getSimbolSpanFromStatus(testCount));
                containerDiv.appendChild(ancestorLink);
                containerDiv.appendChild(testCountSpan);
            });
        });
        return elements;
    }
    static getSimbolSpanFromStatus({ testTotal, passingTests, pendingTests, failingTests }) {
        const span = document.createElement("span");
        span.classList.add("summary-test-label", "test", "smallLabel");
        if (failingTests > 0) {
            span.textContent = "X";
            span.classList.add("fail");
        }
        else if (pendingTests === testTotal) {
            span.textContent = "O";
            span.classList.add("pending");
        }
        else if (passingTests + pendingTests === testTotal) {
            span.textContent = "✓";
            span.classList.add("pass");
        }
        return span;
    }
}
exports.TestSummary = TestSummary;
TestSummary.JOIN_CHAR = ".";
