// bespoken-jest-stare/lib/utils/GenerateInlineReport.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@parcel/core");
const Constants_1 = require("../processor/Constants");
const IO_1 = require("./IO");
const Dependencies_1 = require("../processor/Dependencies");
const path = require("path");
const resultDir = "web/temp/";
const obtainWebFile = (name) => {
    return IO_1.IO.readFileSync(path.resolve(__dirname, "./../../web", name));
};
const obtainJsRenderFile = (name) => {
    return IO_1.IO.readFileSync(path.resolve(__dirname, "./../render/", name));
};
const addThirdParty = (dependency) => {
    const location = require.resolve(dependency.requireDir + dependency.file);
    IO_1.IO.writeFileSync(dependency.targetDir + dependency.file, IO_1.IO.readFileSync(location));
};
IO_1.IO.mkdirsSync(resultDir);
IO_1.IO.copyFileSync("web/template-parcel.html", resultDir + "template.html");
const cssDir = resultDir + Constants_1.Constants.CSS_DIR;
IO_1.IO.mkdirsSync(cssDir);
IO_1.IO.writeFileSync(cssDir + Constants_1.Constants.JEST_STARE_CSS, obtainWebFile(Constants_1.Constants.JEST_STARE_CSS));
const jsDir = resultDir + Constants_1.Constants.JS_DIR;
IO_1.IO.mkdirsSync(jsDir);
IO_1.IO.writeFileSync(jsDir + Constants_1.Constants.JEST_STARE_JS, obtainJsRenderFile(Constants_1.Constants.JEST_STARE_JS));
Dependencies_1.Dependencies.THIRD_PARTY_DEPENDENCIES.forEach((dependency) => {
    const updatedDependency = Object.assign({}, ...[dependency]);
    updatedDependency.targetDir = resultDir + dependency.targetDir;
    addThirdParty(updatedDependency);
});
const bundler = new core_1.default({
    entries: resultDir + "template.html",
    defaultConfig: '@parcel/config-default',
    mode: 'production',
    targets: {
        legacy: {
            engines: {
                "browsers": "> 0.5%, last 2 versions, not dead"
            },
            distDir: "web/dist",
        },
    },
});
try {
    bundler.run().then(({ bundleGraph, buildTime }) => {
        try {
            IO_1.IO.copyFileSync("web/dist/template.html", "web/templateInlineSource.html");
            IO_1.IO.deleteFolderSync("web/dist");
            IO_1.IO.deleteFolderSync("web/temp");
        }
        catch (error) {
            console.log("error while parce bundle", error);
            throw error;
        }
    });
}
catch (error) {
    console.log("error while parce bundle", error);
    throw error;
}
